crates/scroll/hardforks/src/hardfork.rs

```diff
@@ -0,0 +1,264 @@
+//! Hard forks of scroll protocol.
+
+use alloc::{boxed::Box, format, string::String, vec};
+use core::{
+    any::Any,
+    fmt::{self, Display, Formatter},
+    str::FromStr,
+};
+
+use alloy_chains::{Chain, NamedChain};
+use reth_ethereum_forks::{hardfork, ChainHardforks, EthereumHardfork, ForkCondition, Hardfork};
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+hardfork!(
+    /// The name of the Scroll hardfork
+    ///
+    /// When building a list of hardforks for a chain, it's still expected to mix with
+    /// [`EthereumHardfork`].
+    ScrollHardfork {
+        /// Archimedes switch block (nil = no fork, 0 = already on archimedes)
+        Archimedes,
+        /// Bernoulli: <https://scroll.io/blog/blobs-are-here-scrolls-bernoulli-upgrade>.
+        Bernoulli,
+        /// Curie: <https://scroll.io/blog/compressing-the-gas-scrolls-curie-upgrade>.
+        Curie,
+        /// Darwin: <https://scroll.io/blog/proof-recursion-scrolls-darwin-upgrade>.
+        Darwin,
+        /// DarwinV2 <https://x.com/Scroll_ZKP/status/1830565514755584269>.
+        DarwinV2,
+    }
+);
+
+impl ScrollHardfork {
+    /// Retrieves the activation block for the specified hardfork on the given chain.
+    pub fn activation_block<H: Hardfork>(self, fork: H, chain: Chain) -> Option<u64> {
+        // will replace scroll_sepolia after alloy-chains release new tag
+        if chain == Chain::from_named(NamedChain::ScrollSepolia) {
+            return Self::scroll_sepolia_activation_block(fork);
+        }
+        if chain == Chain::from_named(NamedChain::Scroll) {
+            return Self::scroll_mainnet_activation_block(fork);
+        }
+
+        None
+    }
+
+    /// Retrieves the activation timestamp for the specified hardfork on the given chain.
+    pub fn activation_timestamp<H: Hardfork>(self, fork: H, chain: Chain) -> Option<u64> {
+        if chain == Chain::from_named(NamedChain::ScrollSepolia) {
+            return Self::scroll_sepolia_activation_timestamp(fork);
+        }
+        if chain == Chain::from_named(NamedChain::Scroll) {
+            return Self::scroll_mainnet_activation_timestamp(fork);
+        }
+
+        None
+    }
+
+    /// Retrieves the activation block for the specified hardfork on the Scroll Sepolia testnet.
+    pub fn scroll_sepolia_activation_block<H: Hardfork>(fork: H) -> Option<u64> {
+        match_hardfork(
+            fork,
+            |fork| match fork {
+                EthereumHardfork::Homestead |
+                EthereumHardfork::Tangerine |
+                EthereumHardfork::SpuriousDragon |
+                EthereumHardfork::Byzantium |
+                EthereumHardfork::Constantinople |
+                EthereumHardfork::Petersburg |
+                EthereumHardfork::Istanbul |
+                EthereumHardfork::Berlin |
+                EthereumHardfork::London |
```

Comment:
l2geth enable this https://github.com/scroll-tech/testnet/blob/mainnet/core/genesis/genesis.mainnet.json#L14

Reply:
Try to understand your mean: Your mean current genesis file don't exist the baseFee parameter, so the `base_fee_params` is None. right? If yes, I think it will have the value after implemented the [BaseFeeFormula](https://github.com/scroll-tech/reth/issues/60)

Reply:
I walkthrough the code found that the baseFee calculated from the base_fee_params

Reply:
after discussed with @greged93, because Scroll mainnet supported EIP1559 from curie, but the EthChainSpec::genesis_header calculate the EIP1559 `base_fee_per_gas` [from genesis](https://github.com/scroll-tech/reth/blob/66b7e0016383bbfba322a463295d4a65c70fc4cf/crates/chainspec/src/spec.rs#L319), and our mainnet genesis base_fee_per_gas is 0, so need to be compatible with this logic

Reply:
yes, I think you need to keep London fork but implement custom logic for the genesis

Reply:
I fixed like this [c6ca5aa](https://github.com/scroll-tech/reth/pull/56/commits/c6ca5aaa4b93c4871985ce1a353d6290d1bfc97d), pls take a look